src/harmonic/electron_h0_wannier.cpp

```diff
@@ -375,4 +497,435 @@ void ElectronH0Wannier::addShiftedVectors(Eigen::Tensor<double,3> degeneracyShif
+        KOKKOS_LAMBDA(int iK, int iR) {
+          double arg = 0.0;
+          for (int i = 0; i < 3; i++) {
+            arg += cartesianCoordinates(iK, i) * bravaisVectors_d(iR, i);
+          }
+          elPhases_d(iK, iR) = exp(complexI * arg) / vectorsDegeneracies_d(iR);
+        });
+
+    Kokkos::parallel_for(
+        "el_hamilton", Range3D({0, 0, 0}, {numK, numWannier, numWannier}),
+        KOKKOS_LAMBDA(int iK, int m, int n) {
+          Kokkos::complex<double> tmp(0.0);
+          for (int iR = 0; iR < numVectors; iR++) {
+            tmp += elPhases_d(iK, iR) * h0R_d(m, n, iR);
+          }
+          hamiltonians(iK, m, n) = tmp;
+        });
+    Kokkos::realloc(elPhases_d, 0, 0);
+
+  } else {
+    auto vectorsShifts_d = this->vectorsShifts_d;
+    auto degeneracyShifts_d = this->degeneracyShifts_d;
+
+    Kokkos::parallel_for(
+        "elHamiltonianShifted_d",
+        Range3D({0, 0, 0}, {numK, numWannier, numWannier}),
+        KOKKOS_LAMBDA(int iK, int iw1, int iw2) {
+          Kokkos::complex<double> tmp(0.0);
+          for (int iR = 0; iR < numVectors; iR++) {
+            for (int iDeg = 0; iDeg < degeneracyShifts_d(iw1, iw2, iR); ++iDeg) {
+              double arg = 0.;
+              for (int i = 0; i < 3; ++i) {
+                arg += cartesianCoordinates(iK, i) * vectorsShifts_d(i, iDeg, iw1, iw2, iR);
+              }
+              Kokkos::complex<double> phase = exp(complexI * arg)
+                  / vectorsDegeneracies_d(iR) * degeneracyShifts_d(iw1, iw2, iR);
+              tmp += h0R_d(iR, iw1, iw2) * phase;
+            }
+          }
+          hamiltonians(iK, iw1, iw2) = tmp;
+        });
+  }
+
+  return hamiltonians;
+}
+
+/**
+ * Create and diagonalize Hamiltonians for a batch of k-points
+ */
+std::tuple<DoubleView2D, StridedComplexView3D> ElectronH0Wannier::kokkosBatchedDiagonalizeFromCoordinates(
+    const DoubleView2D &cartesianCoordinates, const bool withMassScaling) {
+
+  int numWannier = this->numWannier; // Kokkos quirkyness
+
+  // build Hamiltonians
+  StridedComplexView3D blochHamiltonians =
+      kokkosBatchedBuildBlochHamiltonian(cartesianCoordinates);
+
+  // now the diagonalization.
+
+  int numK = blochHamiltonians.extent(0);
+  DoubleView2D allEnergies("energies_d", numK, numWannier);
+
+  // perform diagonalization
+  kokkosZHEEV(blochHamiltonians, allEnergies);
+  // blochHamiltonians now contains eigenvectors
+
+  return std::make_tuple(allEnergies, blochHamiltonians);
+}
+
+/**
+ * Build and diagonalize Hamiltonians, with velocities
+ */
+std::tuple<DoubleView2D, StridedComplexView3D, ComplexView4D>
+ElectronH0Wannier::kokkosBatchedDiagonalizeWithVelocities(
+    const DoubleView2D &cartesianCoordinates) {
+
+  int numWannier = this->numWannier; // Kokkos quirkyness
+
+  int numK = cartesianCoordinates.extent(0);
+
+  double delta = 1.0e-8;
+  double threshold = 0.000001 / energyRyToEv;// = 1 micro-eV
+
+  // prepare all the wavevectors at which we need the hamiltonian
```

Comment:
@anjohan Again maybe a comment about indexing (if it's not exactly the same as somewhere else in this part of the code)?